feat: compact_object_store_chunks : initial implementation that also lists steps to complete

NGA-TRAN opened the pull request:
This is the first part of  #3089. The function is available but only step 1 is implemented and tested. Step 2-4 will be in the following PRs

- [x] I've read the contributing section of the project [README](https://github.com/influxdata/influxdb_iox/blob/main/README.md).
- [x] Signed [CLA](https://influxdata.com/community/cla/) (if not already signed).

NGA-TRAN commented on server/src/db.rs:
```diff
@@ -651,6 +651,40 @@ impl Db {
+            let partition = partition.read();
+
+            // todo: set these chunks
+            let chunks = vec![];
```
This will be done after all todos inside `compact_object_store_chunks(partition, chunks)` completed

crepererum commented on lifecycle/src/policy.rs:
```diff
@@ -908,6 +909,20 @@ mod tests {
+            partition: LifecycleWriteGuard<'_, TestPartition, Self>,
+            chunks: Vec<LifecycleWriteGuard<'_, TestChunk, Self::Chunk>>,
+        ) -> Result<TaskTracker<()>, Self::Error> {
+            // This is just a trait function that this test may not need
```
if you don't need it, just put `unimplemented!("some reason")` in there.

crepererum commented on server/src/db/catalog/partition.rs:
```diff
@@ -368,6 +372,23 @@ impl Partition {
 
+    /// Return true if there are no other persisted chunks that are in the middle of
+    /// the provided chunk orders
+    pub fn contiguous_object_store_chunks(&self, chunk_orders: &BTreeSet<ChunkOrder>) -> bool {
```
This function is complex enough to demand a unit test.

crepererum commented on server/src/db/catalog/partition.rs:
```diff
@@ -368,6 +372,23 @@ impl Partition {
+            if chunk.is_persisted() {
+                let order = chunk.order();
+                // this chunk does not belong to chunk_orders but in the middle of them
+                if !chunk_orders.contains(&order) && chunk_orders.range(order..).count() > 0 {
```
`chunk_orders.range(order..).count() > 0` seems like a really costly check just to determine if there's anything at/after `order` within `chunk_order`.

You could probably get the last element once (before iterating over `chunks`):

```rust
let last_element: Option<ChunkOrder> = chunk_orders.iter().rev().next();
```

and then just compare that:

```rust
last_element.map(|o| order <= o).unwrap_or(false)
```

crepererum commented on server/src/db/lifecycle/compact_object_store.rs:
```diff
@@ -0,0 +1,500 @@
+    partition: LifecycleWriteGuard<'_, Partition, LockableCatalogPartition>,
+    chunks: Vec<LifecycleWriteGuard<'_, CatalogChunk, LockableCatalogChunk>>,
+    registration: &TaskRegistration,
+) -> Result<(
+    Time,
+    Time,
+    u64,
+    HashSet<Arc<DeletePredicate>>,
+    Vec<Arc<DbChunk>>,
+    ChunkOrder,
+    DatabaseCheckpoint,
+    PartitionCheckpoint,
+)> {
```
Please use a struct here, esp. since the tuple elements are not strongly-typed (e.g. `u64` could be anything). Always prefer a struct over a tuple + docs. Performance-wise they are the same btw.

crepererum commented on server/src/db/lifecycle/compact_object_store.rs:
```diff
@@ -0,0 +1,500 @@
+
+            // Set chunk in the right action which is compacting object store
+            // This function will also error out if the chunk is not yet persisted
+            chunk.set_compacting_object_store(registration)?;
```
I'm pretty sure we have this bug in other places as well, but whenever this state transition fails (e.g. at the third chunk) you might have some chunks already in the new lifecycle (e.g. the first two chunks). This lifecycle action will never be cleaned and the chunks are in this state forever. So ideally we would abort the lifecycle action here as well.

NGA-TRAN replied:
This is a good point. I think this action is needed to let other conflicting processes not change these chunks. However, we have not tried to change the action back after we are done. In this case, if it succeeds, all chunks will be dropped and we do not need to do anything. However, if it fails, then we have to set the action to the chunks' original ones.

What do you think if I create a ticket to handle setting the action to its stage before compacting/persisting for not only this case but for all available code that we believe hitting this bug? @crepererum and @tustvold

tustvold replied:
There is already code to handle this - https://github.com/influxdata/influxdb_iox/blob/main/lifecycle/src/policy.rs#L412

There has been discussion around removing the need for this loop as a performance optimisation, but it hasn't been a priority.

NGA-TRAN replied:
> removing the need for this loop as a performance optimisation

@tustvold : Can you be clearer what `this loop` means? Do I need any action here?

tustvold replied:
> `this loop`

The code loops through every chunk looking for failed lifecycle actions, it would be nice to detect the failure as a consequence of some handle being dropped and enqueue recovery at "source". This would eliminate the need to go looking for failed jobs. The main reason it was implemented this way is to allow for backoff, but backoff could be achieved in other ways.

>  Do I need any action here

No action needed, if it shows up in a profile we can look into then :smile:

NGA-TRAN replied:
Sounds good. In this specific case we need this loop to check for OS chunks and get their chunk order anyway.

crepererum commented on server/src/db/lifecycle/compact_object_store.rs:
```diff
@@ -0,0 +1,500 @@
+    let table_name = Arc::from(partition_addr.table_name.to_string());
+    let partition_key = Arc::from(partition_addr.partition_key.to_string());
+
+    let chunk_id = ChunkId::new();
```
@tustvold Since we've talked about it: it's perfectly fine to create freestanding chunk IDs w/o involving the in-mem catalog.

NGA-TRAN replied:
Any action for me here or this message is for @tustvold only?

crepererum replied:
This was only meant to be an FYI, nothing actionable :)

crepererum commented on server/src/db/lifecycle/compact_object_store.rs:
```diff
@@ -0,0 +1,500 @@
+        let chunk3 = chunks[2].read();
+        let compact_non_persisted_chunks =
+            compact_object_store_chunks(partition, vec![chunk1.upgrade(), chunk3.upgrade()]);
+        assert!(compact_non_persisted_chunks.is_err());
```
please check the error variant to string repr.

crepererum commented on server/src/db/lifecycle/compact_object_store.rs:
```diff
@@ -0,0 +1,500 @@
+        let compact_no_chunks = compact_object_store_chunks(partition, vec![]);
+        assert!(compact_no_chunks.is_err());
+
+        // test 2: persisted non persisted chunks
```
Could you please split this into multiple test functions? I find "series-style" tests quite hard to reason about.

crepererum reviewed (commented):
Needs style and test changes, but I think overall that's a good start from where we can piece-by-piece add the remaining functionality + tests (like partition/db checkpoints and delete predicates, ...).

NGA-TRAN commented on server/src/db/lifecycle/compact_object_store.rs:
```diff
@@ -0,0 +1,500 @@
+    time_of_last_write: Time,
+    chunk_order: ChunkOrder,
+) -> Result<Arc<ParquetChunk>> {
+    // Todo: ask Marco if this cleanup_lock is needed
```
@crepererum See question above

crepererum replied:
Yes, and I think you're not holding it long enough here. The lock MUST be acquired BEFORE you store a parquet file and MUST be released AFTER you completed the preserved catalog transaction. Otherwise the "cleanup parquet files"-background-worker might just delete the parquet file that you are about to add to the catalog.

NGA-TRAN commented:
@crepererum I have addressed all you comments.
@tustvold : you have to review this first thing Monday morning or I will bug you 😄 
I need this PR in to have the following pieces go easier

crepererum reviewed (changes requested):
The DB clean-up lock needs to be moved a bit.

tustvold commented on server/src/db/lifecycle/compact_object_store.rs:
```diff
@@ -0,0 +1,572 @@
+        time::Duration,
+    };
+
+    // Todo: this as copied from persist.rs and should be revived to match the needs here
```
I think you can just use `make_db` here which calls `TestDb::builder().build().await`.

These tests don't need to be overriding the late_arrive_window_seconds as compacting object store is not concerned with such concepts. By extension I think it doesn't need the time provider, as Wall time has no bearing on its behaviour. 

`worker_backoff_millis` is just a hack to stop the lifecycle policy interfering and starting compactions, etc... in parallel - but I think for these tests there is unlikely to be anything to trigger such behaviour and so you might get away without it

tustvold commented on server/src/db/lifecycle/compact_object_store.rs:
```diff
@@ -0,0 +1,572 @@
+        let (db, time) = test_db().await;
+        let late_arrival = Duration::from_secs(1);
+        write_lp(db.as_ref(), "cpu,tag1=cupcakes bar=1 10").await;
+        time.inc(late_arrival);
```
All this stuff about time providers and late arrival is not needed here, as we aren't concerned with persisting unpersisted data (which must have been in-memory for late_arrival_seconds)

tustvold commented on server/src/db/lifecycle/compact_object_store.rs:
```diff
@@ -0,0 +1,572 @@
+        //
+        // Add a MUB
+        write_lp(db.as_ref(), "cpu,tag1=chunk4,tag2=a bar=2 40").await;
+        // todo: Need to ask Marco why there is no handle created here
```
Legacy comment?

tustvold commented on server/src/db/lifecycle/compact_object_store.rs:
```diff
@@ -0,0 +1,572 @@
+        let err = compact_non_contiguous_persisted_chunks.unwrap_err();
+        assert!(err
+            .to_string()
+            .contains("Cannot compact the provided persisted chunks. They are not contiguous"));
```
Would be good to print the error string on assertion failure

tustvold reviewed (commented):
Looks good to me, some minor nits.

tustvold commented on server/src/db/lifecycle/compact_object_store.rs:
```diff
@@ -0,0 +1,572 @@
+    let time_of_last_write = time_of_last_write.expect("Should have had a last write somewhere");
+
+    // drop partition lock
+    let _partition = partition.into_data().partition;
```
Could also use `std::mem::drop(partition)` or simply leave it to exiting the scope of the function to drop the lock

crepererum commented on server/src/db/lifecycle/compact_object_store.rs:
```diff
@@ -82,26 +82,39 @@ pub(crate) fn compact_object_store_chunks(
+        // Step 3: Start to persist files and update the preserved catalog accordingly
+        // This process needs to hold cleanup lock to avoid the persisted file was deleted right after
+        // it is created and before it is updated in the preserved catalog
+        {
```
:+1:

NGA-TRAN commented:
@tustvold 
Hah: I forgot to remove "ataumerge" and kodiak has merged it after it got an approval. I am addressing your comments and they will go with my next PR which might include new pieces

NGA-TRAN reviewed (commented):
ALl @tustvold 's comments have been addressed in #3183
